src/integrationTest/java/org/opensearch/security/http/AsyncTests.java

```diff
@@ -0,0 +1,115 @@
+        .authc(AUTHC_HTTPBASIC_INTERNAL)
+        .users(ADMIN_USER)
+        .rolesMapping(new RolesMapping(ALL_ACCESS).backendRoles("admin"))
+        .anonymousAuth(false)
+        .nodeSettings(Map.of(ConfigConstants.SECURITY_RESTAPI_ROLES_ENABLED, List.of(ALL_ACCESS.getName())))
+        .build();
+
+    @Test
+    public void testBulkAndCacheInvalidationMixed() throws Exception {
+        String indexName = "test-index";
+        final String invalidateCachePath = "/_plugins/_security/api/cache";
+        final String nodesPath = "/_nodes";
+        final String bulkPath = "/_bulk";
+        final String document = ("{ \"index\": { \"_index\": \"" + indexName + "\" }}\n{ \"foo\": \"bar\" }\n").repeat(5);
+        final int parallelism = 5;
+        final int totalNumberOfRequests = 30;
+
+        try (final TestRestClient client = cluster.getRestClient()) {
+            IndexOperationsHelper.createIndex(cluster, indexName);
+
+            final CountDownLatch countDownLatch = new CountDownLatch(1);
+
+            final var cacheInvalidationRequests = AsyncActions.generate(() -> {
+                countDownLatch.await();
+                System.err.println("Generation triggered invalidate cache requests");
+                final HttpDelete delete = new HttpDelete(client.getHttpServerUri() + invalidateCachePath);
+                return client.executeRequest(delete, getBasicAuthHeader(ADMIN_USER.getName(), ADMIN_USER.getPassword()));
+            }, parallelism, totalNumberOfRequests);
+
+            final var bulkRequests = AsyncActions.generate(() -> {
+                countDownLatch.await();
+                System.err.println("Generation triggered bulk requests");
+                final HttpPost post = new HttpPost(client.getHttpServerUri() + bulkPath);
+                post.setEntity(new ByteArrayEntity(document.getBytes(StandardCharset.UTF_8), ContentType.APPLICATION_JSON));
+                return client.executeRequest(post, getBasicAuthHeader(ADMIN_USER.getName(), ADMIN_USER.getPassword()));
+            }, parallelism, totalNumberOfRequests);
+
+            final var nodesRequests = AsyncActions.generate(() -> {
+                countDownLatch.await();
+                System.err.println("Generation triggered node requests");
+                final HttpGet get = new HttpGet(client.getHttpServerUri() + nodesPath);
+                return client.executeRequest(get, getBasicAuthHeader(ADMIN_USER.getName(), ADMIN_USER.getPassword()));
+            }, parallelism, totalNumberOfRequests);
+
+            // Make sure all requests start at the same time
+            countDownLatch.countDown();
+
+            List<CompletableFuture<HttpResponse>> allRequests = new ArrayList<CompletableFuture<HttpResponse>>();
+            allRequests.addAll(cacheInvalidationRequests);
+            allRequests.addAll(bulkRequests);
+            allRequests.addAll(nodesRequests);
+            Collections.shuffle(allRequests);
```

Comment:
Nit: Unneeded

Reply:
Won't the order of the requests play some role in how they are executed? Once a thread is free to take new work, it won't just take a random item from the list. It will take them in order so the order should be random.

Reply:
The ordering in the output collection is not related to the order of execution.  `CompletableFuture.supplyAsync(...)` schedules all the tasks in the executor supplied.  In this test case you have 3 _different_ executors that can run 5 actions at a time.  By using the count down latch you force all of the tasks to wait to execute until you've set everything up.  

There is no way to 'know' deterministically what will happen after the latch triggers because those IO operations can queue/delay/run based on what java + the operation system decide.  The order of those responses is just the order the tasks were initially queued in.  For this test that ordering/grouping doesn't matter cause as all responses should be 200s.

Reply:
Got it. Thanks for the explanation. I removed the shuffle.